File: apps/android/app/src/main/java/hub/core/android/chat/ToolActivity.kt
package hub.core.android.chat

import hub.core.client.model.ToolCall
import hub.core.client.model.ToolCallStatus

/** The folded row of a finished turn's tools: how many, how long, how many failed, which. */
data class ToolActivitySummary(
    val count: Int,
    val failed: Int,
    /** Wall-clock time from the first start to the last finish, else the calls' durations summed. */
    val durationMs: Long?,
    /** The latest distinct tool names, most recent first. */
    val names: List<String>,
)

data class ToolActivityState(
    /** True once the turn has ended and nothing is still running or waiting. */
    val folded: Boolean,
    /** The calls a live turn shows, in order; empty when folded (the row opens to all of them). */
    val visible: List<ToolCall>,
    /** How many calls the "+k earlier steps" line (live) or the folded row stands for. */
    val hidden: Int,
    val summary: ToolActivitySummary,
)

/**
 * What a turn's tool calls show — the web's `toolActivity` (packages/web/src/chat/toolActivity.ts)
 * with the phone's window of two (owner, 2026-09-26; DECISIONS §111):
 *
 * - while the turn runs, the last [WINDOW] calls, plus any call still running or waiting and any
 *   that failed (an error stays in view until the turn ends); the rest are "+k earlier steps";
 * - once it has ended, every call folds into one summary row that opens to the full list.
 */
object ToolActivity {
    const val WINDOW = 2
    const val NAMES = 2

    private fun ToolCall.busy() = status == ToolCallStatus.RUNNING || status == ToolCallStatus.AWAITING_APPROVAL
    private fun ToolCall.pinned() = busy() || status == ToolCallStatus.FAILED

    fun of(calls: List<ToolCall>, live: Boolean, window: Int = WINDOW, maxNames: Int = NAMES): ToolActivityState {
        val summary = summarize(calls, maxNames)
        val folded = !live && calls.none { it.busy() }
        if (folded) return ToolActivityState(true, emptyList(), calls.size, summary)
        val start = (calls.size - window).coerceAtLeast(0)
        val visible = calls.filterIndexed { index, call -> index >= start || call.pinned() }
        return ToolActivityState(false, visible, calls.size - visible.size, summary)
    }

    fun summarize(calls: List<ToolCall>, maxNames: Int = NAMES): ToolActivitySummary {
        val names = mutableListOf<String>()
        for (call in calls.asReversed()) {
            if (names.size >= maxNames) break
            if (call.name !in names) names += call.name
        }
        return ToolActivitySummary(
            count = calls.size,
            failed = calls.count { it.status == ToolCallStatus.FAILED },
            durationMs = totalDuration(calls),
            names = names,
        )
    }

    private fun totalDuration(calls: List<ToolCall>): Long? {
        if (calls.isNotEmpty() && calls.all { it.startedAt != null && it.finishedAt != null }) {
            val start = calls.minOf { it.startedAt!!.toInstant().toEpochMilli() }
            val end = calls.maxOf { it.finishedAt!!.toInstant().toEpochMilli() }
            return (end - start).coerceAtLeast(0)
        }
        val durations = calls.mapNotNull { it.durationMs?.toLong() }
        return if (durations.isEmpty()) null else durations.sum()
    }

    /** Minutes and seconds of the folded row's time, never below one second ("1m 05s", "42s"). */
    fun durationParts(ms: Long): Pair<Long, Long> {
        val total = ((ms + 500) / 1000).coerceAtLeast(1)
        return total / 60 to total % 60
    }
}

File: apps/ios/CoreHub/Chat/ToolActivity.swift
// What a turn's tool calls show — the web's `toolActivity` (packages/web/src/chat/toolActivity.ts)
// and Android's `ToolActivity`, with the phone's window of two (owner, 2026-09-26; DECISIONS §111):
//
// - while the turn runs, the last `window` calls, plus any call still running or waiting and any
//   that failed (an error stays in view until the turn ends); the rest are "+k earlier steps";
// - once it has ended, every call folds into one summary row that opens to the full list.
import CoreHubClient
import Foundation

struct ToolActivitySummary: Equatable {
    var count: Int
    var failed: Int
    /// Wall-clock time from the first start to the last finish, else the calls' durations summed.
    var durationMs: Int?
    /// The latest distinct tool names, most recent first.
    var names: [String]
}

struct ToolActivity: Equatable {
    static let window = 2
    static let names = 2

    /// True once the turn has ended and nothing is still running or waiting.
    var folded: Bool
    /// The calls a live turn shows, in order; empty when folded (the row opens to all of them).
    var visible: [ToolCall]
    /// How many calls the "+k earlier steps" line (live) or the folded row stands for.
    var hidden: Int
    var summary: ToolActivitySummary

    private static func busy(_ call: ToolCall) -> Bool {
        call.status == .running || call.status == .awaitingApproval
    }

    static func of(_ calls: [ToolCall], live: Bool, window: Int = ToolActivity.window, maxNames: Int = ToolActivity.names) -> ToolActivity {
        let summary = summarize(calls, maxNames: maxNames)
        let folded = !live && !calls.contains(where: busy)
        if folded { return ToolActivity(folded: true, visible: [], hidden: calls.count, summary: summary) }
        let start = max(0, calls.count - window)
        let visible = calls.enumerated()
            .filter { $0.offset >= start || busy($0.element) || $0.element.status == .failed }
            .map(\.element)
        return ToolActivity(folded: false, visible: visible, hidden: calls.count - visible.count, summary: summary)
    }

    static func summarize(_ calls: [ToolCall], maxNames: Int = ToolActivity.names) -> ToolActivitySummary {
        var names: [String] = []
        for call in calls.reversed() where names.count < maxNames && !names.contains(call.name) {
            names.append(call.name)
        }
        return ToolActivitySummary(
            count: calls.count,
            failed: calls.filter { $0.status == .failed }.count,
            durationMs: totalDuration(calls),
            names: names
        )
    }

    private static func totalDuration(_ calls: [ToolCall]) -> Int? {
        let starts = calls.compactMap(\.startedAt)
        let ends = calls.compactMap(\.finishedAt)
        if !calls.isEmpty, starts.count == calls.count, ends.count == calls.count,
           let first = starts.min(), let last = ends.max() {
            return max(0, Int((last.timeIntervalSince(first) * 1000).rounded()))
        }
        let durations = calls.compactMap(\.durationMs)
        return durations.isEmpty ? nil : durations.reduce(0, +)
    }

    /// Minutes and seconds of the folded row's time, never below one second ("1m 05s", "42s").
    static func durationParts(_ ms: Int) -> (minutes: Int, seconds: Int) {
        let total = max(1, Int((Double(ms) / 1000).rounded()))
        return (total / 60, total % 60)
    }
}

/// The CLDR plural category of `count` in the app's languages: Arabic has six, English two.
/// The catalogue keeps one key per category (`tool.activity.steps.few`), as the web's does.
enum PluralCategory {
    static func of(_ count: Int, _ language: AppLanguage) -> String {
        switch language {
        case .en:
            return count == 1 ? "one" : "other"
        case .ar:
            let rest = count % 100
            switch count {
            case 0: return "zero"
            case 1: return "one"
            case 2: return "two"
            default:
                if (3...10).contains(rest) { return "few" }
                if (11...99).contains(rest) { return "many" }
                return "other"
            }
        }
    }
}

extension L10n {
    /// `key.<category>` for `count`, with `{count}` filled in.
    func plural(_ key: String, _ count: Int) -> String {
        t("\(key).\(PluralCategory.of(count, language))", ["count": String(count)])
    }
}
